pkg/cdi/annotations_test.go

```diff
@@ -0,0 +1,458 @@
+	}
+
+	for _, tc := range []*testCase{
+		{
+			name:    "invalid, empty plugin",
+			plugin:  "",
+			invalid: true,
+		},
+		{
+			name:    "invalid, empty device ID",
+			plugin:  "plugin",
+			devID:   "",
+			invalid: true,
+		},
+		{
+			name:    "invalid, non-alphanumeric first character",
+			plugin:  "_vendor.class",
+			devID:   "device",
+			invalid: true,
+		},
+		{
+			name:    "invalid, non-alphanumeric last character",
+			plugin:  "vendor.class",
+			devID:   "device_",
+			invalid: true,
+		},
+		{
+			name:    "invalid, plugin contains invalid characters",
+			plugin:  "ven.dor-cl+ass",
+			devID:   "device",
+			invalid: true,
+		},
+		{
+			name:    "invalid, devID contains invalid characters",
+			plugin:  "vendor.class",
+			devID:   "dev+ice",
+			invalid: true,
+		},
+		{
+			name:    "invalid, too plugin long",
+			plugin:  "123456789012345678901234567890123456789012345678901234567",
+			devID:   "device",
+			invalid: true,
+		},
+		{
+			name:   "valid, simple",
+			plugin: "vendor.class",
+			devID:  "device",
+			key:    AnnotationPrefix + "vendor.class" + "_" + "device",
+		},
+		{
+			name:   "valid, with special characters",
+			plugin: "v-e.n_d.or.cl-as_s",
+			devID:  "d_e-v-i-c_e",
+			key:    AnnotationPrefix + "v-e.n_d.or.cl-as_s" + "_" + "d_e-v-i-c_e",
+		},
+		{
+			name:   "valid, with /'s replaced in devID",
+			plugin: "v-e.n_d.or.cl-as_s",
+			devID:  "d-e/v/i/c-e",
+			key:    AnnotationPrefix + "v-e.n_d.or.cl-as_s" + "_" + "d-e_v_i_c-e",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			key, err := AnnotationKey(tc.plugin, tc.devID)
+			if !tc.invalid {
+				require.NoError(t, err, "annotation key")
+				require.Equal(t, tc.key, key, "annotation key")
+			} else {
+				require.Error(t, err)
+			}
+		})
+	}
+}
+
+func TestAnnotateInjection(t *testing.T) {
+	type inject = struct {
+		plugin  string
+		devID   string
+		devices []string
+	}
+	type testCase = struct {
+		name        string
+		existing    map[string]string
+		injections  []*inject
+		annotations map[string]string
+		parsed      []string
+		invalid     bool
+	}
+
+	for _, tc := range []*testCase{
+		{
+			name: "one plugin, one device",
+			injections: []*inject{
+				{
+					plugin: "vendor.class",
+					devID:  "device",
+					devices: []string{
+						"vendor.com/class=device",
+					},
+				},
+			},
+			annotations: map[string]string{
+				AnnotationPrefix + "vendor.class_device": "vendor.com/class=device",
+			},
+			parsed: []string{
+				"vendor.com/class=device",
+			},
+		},
+		{
+			name: "one plugin, multiple devices",
+			injections: []*inject{
+				{
+					plugin: "vendor.class",
+					devID:  "device",
+					devices: []string{
+						"vendor.com/class=device1",
+						"vendor.com/class=device2",
+						"vendor.com/class=device3",
+					},
+				},
+			},
+			annotations: map[string]string{
+				AnnotationPrefix + "vendor.class_device": "vendor.com/class=device1,vendor.com/class=device2,vendor.com/class=device3",
```

Comment:
Nit: I know it's useful to add `AnnotationPrefix` here, but it would also be nice to see in the tests how the full annotation key will look and as such it may be useful to explicitly add the `cdi.k8s.io/` for the tests.